src/link.ts

```diff
@@ -0,0 +1,73 @@
+import {
+  ExternalOperation,
+  ApolloChainLink,
+  ApolloLink,
+  NextLink,
+  Operation,
+} from './types';
+
+import {
+  parse,
+} from 'graphql/language/parser';
+
+export default class Link {
+  public static chain(links: ApolloLink[]): ApolloChainLink {
+    if ( links.length < 1 ) {
+      throw new Error('Must have at least one link to form a chain');
+    }
+
+    const _links = [...links];
+    return {
+      request: (operation: ExternalOperation) => {
+        const _operation = Link.transformOperation(operation);
+
+        const chainStart = Link.buildLinkChain(_links);
+        return chainStart(_operation);
+
+      },
+    };
+  }
+
+  public static asPromiseWrapper(link) {
+    return {
+      request: Link.toPromise(link),
+    };
+  }
+
+  private static toPromise = (link) => {
+    return (operation: Operation, next?: NextLink) => {
+      const observable = link.request(operation, next);
+      return new Promise((resolve, reject) => {
+        observable.subscribe({
+          next: resolve,
+          error: reject,
+        });
+      });
+    };
+  }
+
+  private static buildLinkChain(links: ApolloLink[]): NextLink {
+    const _links = [...links];
```

Comment:
This way of creating the connections between the links makes a new array for each request. It mirrors the middleware and afterware code from the netowrk interface of `apollo-client`(also in `apollo-fetch`) the `next` function on demand. We could create a closure and keep an index that we increment